### PR TITLE
Allow exporting plan to district index file CSV

### LINKED ISSUE #125: Export to CSV
- Export plan to csv file
- Two columns: block ID, district
- Export from map page (top right corner)

Button like this on map page: (with pop-out for "Download CSV")

### PR DESCRIPTION
**Pull request by kshepard**
## Overview

Allow exporting plan to district index file CSV, which is a format that will allow sharing plans among multiple redistricting platforms.

### Checklist

- [x] Description of PR is in an appropriate section of `CHANGELOG.md` and grouped with similar changes, if possible

### Demo

csv-download-1
csv-download-2
csv-download-3

### Notes

I needed to use the geounit hierarchy on the server side for the first time. I debated between pulling it down and loading it into memory upon startup, or just regenerating it when needed. I chose the latter so as to not affect our start up time and memory footprint, especially since we only need it for this one thing at the moment, and it's very quick to generate. If we begin using this in more places on the server side, a case can be made for loading it on startup and keeping it in memory.

I just used some temporary icons for the download/CSV buttons. It would be good for @alexelash or @jfrankl to take a look and customize this.

## Testing Instructions

- `./scripts/update`
- `./scripts/server`
- Load a map
- Download a CSV
- Perform some basic spot checks of it (ids looks like block ids, district ids look like district ids, number of rows looks right)
- Unfortunately it won't be super easy to fully test this until we have the inverse feature in place (importing a district index file CSV)

Closes #125

### CODE REVIEW

**Comment by jfrankl** [comment]
Kenny, could you please change the label to "Export" and use this icon? [export.svg.zip](https://github.com/PublicMapping/districtbuilder/files/5531985/export.svg.zip)

**Comment by kshepard** [comment]
@jfrankl, all set: 

csv-export-1

**Comment by jfrankl** [comment]
Excellent!

File: src/server/src/districts/entities/geo-unit-topology.entity.ts [hunk 1]
```diff
@@ -111,6 +112,64 @@ function getNode(
   } as GeoUnitHierarchy;
 }
 
+// Groups a topology into a hierarchy of geounits corresponding to a district definition structure.
+// Note: this function, along with getNodeForHierarchy are copy-pasted directly (w/rename) from
+// process-geojson. We will need to fix #179 before we can share such code among projects.
```
Comment (maurizi) [inline src/server/src/districts/entities/geo-unit-topology.entity.ts:117]: Let's make #179 a higher priority, this is getting a little silly.

File: src/client/api.ts [hunk 2]
```diff
@@ -175,3 +176,16 @@ export async function patchProject(
...
   });
 }
+
+export async function exportProjectCsv(id: ProjectId): Promise<void> {
+  return new Promise((resolve, reject) => {
+    apiAxios
+      .get(`/api/projects/${id}/export/csv`)
+      .then(response => {
+        return resolve(
+          saveAs(new Blob([response.data], { type: "text/csv;charset=utf-8" }), `${id}.csv`)
```
Comment (maurizi) [inline src/client/api.ts:186]: For a (hopefully) more informative name, consider using the project `name` here instead of `id`:
```diff
diff --git a/src/client/actions/projectData.ts b/src/client/actions/projectData.ts
index f7297e5..38f836f 100644
--- a/src/client/actions/projectData.ts
+++ b/src/client/actions/projectData.ts
@@ -34,5 +34,5 @@ export const updateDistrictsDefinitionRefetchGeoJsonSuccess = createAction(
 
 export const updateProjectFailed = createAction("Update project failure")();
 
-export const exportCsv = createAction("Export project CSV")<ProjectId>();
+export const exportCsv = createAction("Export project CSV")<IProject>();
 export const exportCsvFailure = createAction("Export project CSV failure")<string>();
diff --git a/src/client/api.ts b/src/client/api.ts
index e1ce6ff..c1df717 100644
--- a/src/client/api.ts
+++ b/src/client/api.ts
@@ -177,13 +177,16 @@ export async function patchProject(
   });
 }
 
-export async function exportProjectCsv(id: ProjectId): Promise<void> {
+export async function exportProjectCsv(project: IProject): Promise<void> {
   return new Promise((resolve, reject) => {
     apiAxios
-      .get(`/api/projects/${id}/export/csv`)
+      .get(`/api/projects/${project.id}/export/csv`)
       .then(response => {
         return resolve(
-          saveAs(new Blob([response.data], { type: "text/csv;charset=utf-8" }), `${id}.csv`)
+          saveAs(
+            new Blob([response.data], { type: "text/csv;charset=utf-8" }),
+            `${project.name}.csv`
+          )
         );
       })
       .catch(error => reject(error.message));
diff --git a/src/client/components/ExportMenu.tsx b/src/client/components/ExportMenu.tsx
index 94729c2..2133b9a 100644
--- a/src/client/components/ExportMenu.tsx
+++ b/src/client/components/ExportMenu.tsx
@@ -2,7 +2,7 @@
 import { jsx, Box } from "theme-ui";
 import { Button as MenuButton, Wrapper, Menu, MenuItem } from "react-aria-menubutton";
 import Icon from "../components/Icon";
-import { ProjectId } from "../../shared/entities";
+import { IProject } from "../../shared/entities";
 import { style, invertStyles } from "./MenuButton.styles";
 import store from "../store";
 import { exportCsv } from "../actions/projectData";
@@ -13,7 +13,7 @@ enum UserMenuKeys {
 
 interface ExportProps {
   readonly invert?: boolean;
-  readonly projectId: ProjectId;
+  readonly project: IProject;
 }
 
 const ExportMenu = (props: ExportProps) => {
@@ -21,7 +21,7 @@ const ExportMenu = (props: ExportProps) => {
     <Wrapper
       sx={{ position: "relative", pr: 1 }}
       onSelection={(userMenuKey: string) =>
-        userMenuKey === UserMenuKeys.ExportCsv && store.dispatch(exportCsv(props.projectId))
+        userMenuKey === UserMenuKeys.ExportCsv && store.dispatch(exportCsv(props.project))
       }
     >
       <MenuButton
diff --git a/src/client/components/ProjectHeader.tsx b/src/client/components/ProjectHeader.tsx
index 62f2b7e..3f12de7 100644
--- a/src/client/components/ProjectHeader.tsx
+++ b/src/client/components/ProjectHeader.tsx
@@ -103,7 +103,7 @@ const ProjectHeader = ({
           )}
           <ShareMenu invert={true} />
           <SupportMenu invert={true} />
-          {project ? <ExportMenu invert={true} projectId={project.id} /> : null}
+          {project ? <ExportMenu invert={true} project={project} /> : null}
           <Box sx={{ position: "relative" }}>
             <Button
               sx={{
```

**Review by maurizi — APPROVED** [review]
I left a few small comment clean up suggestions as well as a larger code suggestion (which I don't feel strongly about, so feel free to disregard).

Otherwise, this looks great! Nice to see that it wasn't too computationally intensive to build that data structure as needed, and I agree that's a better performance trade-off.

File: src/server/src/districts/entities/geo-unit-topology.entity.ts (hunk 1 above)
Comment (kshepard) [reply src/server/src/districts/entities/geo-unit-topology.entity.ts:117]: Yeah, it is a little sad thinking it'll be easy to share some code, and then remembering that it's not quite at that point.

File: src/client/api.ts (hunk 2 above)
Comment (kshepard) [reply src/client/api.ts:186]: Cool, the name does sound like a better choice for a filename, thanks for the suggestion.

**Pull request merged**